src/Admin/Settings/settings.php

```diff
@@ -88,11 +88,22 @@ function sanitize_value( $field_name, $value, $original_value ) {
 			$value = $original_value;
 		}
 
-		if ( ! wp_http_validate_url( $value ) ) {
+		// set a default scheme if none is supplied, otherwise esc_url_raw() doesn't like it.
+		$scheme_added = false;
+		if ( is_null( wp_parse_url( $value, PHP_URL_SCHEME ) ) ) {
+			$value        = 'https://' . $value;
+			$scheme_added = true;
+		}
+
+		if ( esc_url_raw( $value ) !== $value ) {
 			// translators: %s is the value the user entered.
 			add_error( 'homeserver-invalid', sprintf( __( '<tt>%s</tt> is not a valid homeserver URL.' ), $value ) );
 			$value = $original_value;
 		}
+
+		if ( $scheme_added ) {
+			$value = substr( $value, 8 );
```

Comment:
I think we could instead output the prefixed URL, i.e. `https://example.com`, otherwise it might be confusing for the user, since the error message mentions `https://` but the value that shows in the `<input>` doesn't.

Removing the prefix here also prevents the user from editing `https://` to `http://` in case they need to. They would need to guess that they need to input `http://`.

Reply:
I made the change to show the correct value in the error message. Adding scheme is only for validating purposes. I don't want to add scheme and save it since `matrix.org` is a valid homeserver string, and where we connect to for that homeserver happens with well-known resolution which can be anything (same host with http/https or diff host). So, I want to retain that context as is, entered by the user. Makes sense?